File: cs/State.cs
using System;
using System.Collections.Generic;
namespace cs
{
    class CashboxWithState
    {
        private Dictionary<string, Payment> payments= new Dictionary<string, Payment>();
        private string state;
       
        public CashboxWithState()
        {
            payments.Add("CASH", new CashPayment());
            payments.Add("CARD", new CardPayment());
            
        }

        public List<string> States
        {
            get
            {
                List<string> result = new List<string>();
                foreach (string state in this.payments.Keys)
                {
                    result.Add(state);
                }
                return result;
            }
        }    
        public string State {
            get {
                return this.state;
            }
            set{
                if (this.States.Contains(value))
                    this.state = value;
                else
                    throw new Exception("Unknown state");
            }
        }
        public bool TakePayment(double sum)
        {
            Payment payment;
            if (payments.TryGetValue(this.state, out payment))
            {
                return payment.Pay(sum);
            }
            return false;
        }

        
    }
}

File: js/State.js
import promptSync from "prompt-sync";
let prompt = promptSync();

class CashPayment{
    Pay(PaySum){
        let Cash = Number(prompt("Внесіть готівку:"));
        let Change= Cash - PaySum;
        if (Change>0)
        {
            console.log(`Ваша решта ${Change}`);
            return true;
        }
        if (Change==0)
            return true;
        return false;
    }        
}

class CardPayment{
    Pay(PaySum){
        console.log("Оплата каpтою");            
        return true;
    }
}

class CashboxWithState {
    constructor(){
        this.paymentStates = {
            "CASH": new CashPayment(),
            "CARD": new CardPayment(),
        }
        this.state = "CASH";
    }
    
    get States () {
        return Object.getOwnPropertyNames(this.paymentStates); 
    }

    set State(value){
        if (this.States.indexOf(value)>-1)
            this.state=value;
        else
            throw "Unknown state";
    }

    TakePayment(sum){
        if (this.paymentStates[this.state])
            return this.paymentStates[this.state].Pay(sum);
        return false;    
    }

    
}


export default CashboxWithState;